tests/classes/test_types.py

```diff
@@ -0,0 +1,291 @@
+            raw_type(
+                hiku.types.Sequence[hiku.types.Optional[hiku.types.Integer]]
+            ),
+        ),
+        (
+            list[str] | None,
+            raw_type(
+                hiku.types.Optional[hiku.types.Sequence[hiku.types.String]]
+            ),
+        ),
+        (
+            typing.Optional[typing.List[str]],
+            raw_type(
+                hiku.types.Optional[hiku.types.Sequence[hiku.types.String]]
+            ),
+        ),
+        # some absurd stuff just for the sake of it
+        (
+            list[list[bool | None]] | None,
+            raw_type(
+                hiku.types.Optional[
+                    hiku.types.Sequence[
+                        hiku.types.Sequence[
+                            hiku.types.Optional[hiku.types.Boolean]
+                        ]
+                    ]
+                ]
+            ),
+        ),
+        (
+            typing.Optional[typing.List[typing.List[typing.Optional[bool]]]],
+            raw_type(
+                hiku.types.Optional[
+                    hiku.types.Sequence[
+                        hiku.types.Sequence[
+                            hiku.types.Optional[hiku.types.Boolean]
+                        ]
+                    ]
+                ]
+            ),
+        ),
+    ],
+)
+def test_to_hiku_type__scalars(typ, expected):
+    assert to_hiku_type(typ) == expected
+
+
+@pytest.mark.parametrize(
+    "typ",
+    [
+        complex,
+        tuple,
+        typing.Any,
+        None,
+        1,
+        list,
+        typing.Optional,
+        int | str,
+        typing.Union[bool, float],
+        typing.Annotated[int, lazy(".")],
+        ref["Droid"],
+    ],
+)
+def test_to_hiku_type__raises(typ):
+    with pytest.raises(ValueError):
+        to_hiku_type(typ)
+
+
+@node(name="HumanNodeName")
+class Human:
+    __key__: int
+
+    id: int
+
+
+@pytest.mark.parametrize(
+    "typ,expected",
+    [
+        (ref[Human], raw_type(hiku.types.TypeRef["HumanNodeName"])),
+        (
+            ref[Human] | None,
+            raw_type(hiku.types.Optional[hiku.types.TypeRef["HumanNodeName"]]),
+        ),
+        (
+            typing.Optional[ref[Human]],
+            raw_type(hiku.types.Optional[hiku.types.TypeRef["HumanNodeName"]]),
+        ),
+        (
+            list[ref[Human]],
+            raw_type(hiku.types.Sequence[hiku.types.TypeRef["HumanNodeName"]]),
+        ),
+        (
+            typing.List[ref[Human]],
+            raw_type(hiku.types.Sequence[hiku.types.TypeRef["HumanNodeName"]]),
+        ),
+    ],
+)
+def test_to_hiku_type__ref(typ, expected):
+    assert to_hiku_type(typ) == expected
+
+
+@pytest.mark.parametrize(
+    "typ,expected",
+    [
+        (
+            typing.Annotated[ref["Droid"], lazy(".droid")],
+            _LazyTypeRef(
+                classname="Droid",
+                module=".droid",
+                package="tests.classes",
```

Comment:
I don't think there really is a good enough way to do it differently.

As I see it:
1. Using TYPE_CHECKING imports is not a runtime thing, we could process it by parsing sources and so on, but IMO that looks as a bit too much.
2. Using user-controlled imports in a nested function, etc: I don't see a way how to make it work with annotations, that probably requires making fundamentally different API.
3. Selecting objects by name from internal registry (sqlalchemy way). That could work just fine, but again is quite implicit and there still have to be ways to handle name conflicts between classes (generated graphql types may have different names, it is fine). SQLAlchemy basically does the same [thing](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#late-evaluation-of-relationship-arguments) as we do, but uses function args instead of `typing.Annotated`. I am open to make `lazy` param optional in hiku (in case there are no name conflicts), but I would rather implement it later, after the core is done.

Reply:
> I would rather implement it later, after the core is done.

Sounds good for me